_posts/2021-05-04-intro-to-criteria-api.md

```diff
@@ -0,0 +1,473 @@
+          <artifactId>hibernate-jpamodelgen</artifactId>
+          <version>5.3.7.Final</version>
+      </dependency>
+  </dependencies>
+    
+````
+
+````xml   
+  <build>
+      <plugins>
+          <plugin>
+              <groupId>org.springframework.boot</groupId>
+              <artifactId>spring-boot-maven-plugin</artifactId>
+          </plugin>
+          <plugin>
+              <artifactId>maven-compiler-plugin</artifactId>
+              <configuration>
+                  <source>11</source>
+                  <target>11</target>
+                  <compilerArguments>
+                      <processor>org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor</processor>
+                  </compilerArguments>
+              </configuration>
+          </plugin>
+      </plugins>
+  </build>
+```` 
+By building the project, the Metamodel classes will be generated automatically in the `target/generated-classes` folder. 
+**We can access all our entity object fields through the generated classes. Metamodel class name for `Product` entity is `Product_` 
+like the following**: 
+
+````java
+@Generated(value = "org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor")
+@StaticMetamodel(Product.class)
+public abstract class Product_ {
+
+  public static volatile SingularAttribute<Product, Double> price;
+  public static volatile SingularAttribute<Product, String> name;
+  public static volatile SingularAttribute<Product, Long> id;
+  public static volatile SingularAttribute<Product, String> category;
+
+  public static final String PRICE = "price";
+  public static final String NAME = "name";
+  public static final String ID = "id";
+  public static final String CATEGORY = "category";
+
+}
+````
+
+We repeat `findProductsByNameAndCategory` implementation, this time we're using Metamodel classes in the `Predicates` 
+definition as you can see here:
+
+````java
+@Repository
+public class ProductRepositoryImpl {
+
+  @PersistenceContext
+  EntityManager em;
+
+  public List<Product> findProductsByNameAndCategory(
+                                      String name,
+                                      String category) {
+
+    CriteriaBuilder cBuilder = em.getCriteriaBuilder();
+    CriteriaQuery<Product> cQuery = cBuilder.createQuery(Product.class);
+
+    /* FROM **/
+    Root<Product> root = cQuery.from(Product.class);
+
+    /* Conditions using Metamodel **/
+    Predicate namePredicate = cBuilder.
+      like(root.get(Product_.NAME), '%' + (name == null ? "" : name) + '%');
+    Predicate categoryPredicate = cBuilder.
+      equal(root.get(Product_.CATEGORY), (category == null ? Category.TOOLS : category));
+
+    /* WHERE **/
+    cQuery.where(namePredicate, categoryPredicate);
+
+    /* Create Query object **/
+    TypedQuery<Product> query = em.createQuery(cQuery);
+
+    return query.getResultList();
+  }
+}
+````
+**By using Metamodel we can use `Product_.name` and `Product_.category` instead of `"name"` and `"category"`**.
+
```

Comment:
Instead of going into Specifications (see comment below), maybe add a few sections that explain more features of the Criteria API. What are common queries that we would want to do? 

* how can we combine predicates with `AND`?
* how can we combine predicates with `OR`?
* how can we add a `>`, `<`, `>=`, `<=` comparison to a where clause?
* how can we join two tables?
* how can we query for a list of items?
* how can we query for a single item (and what happens if the query returns multiple items)?

After all, this article is about the Criteria API, and should not be about Specifications too much